### DISCUSSION TITLE
App Certificate 및 Provisioning Profile 생성 및 팀원 공유방법 공유

### DISCUSSION THREAD

**Discussion by HanGyeongjun** [Team-ShortcutsZip/MacC-Team-HappyAnding#421] [Announcements]
## Certificate? Provisioning?
1. 아래 블로그 글을 참고해주세요!

[[iOS 앱 배포 준비] 애플에 배포하기 위한 4가지 개념](https://ios-development.tistory.com/246)

[[iOS 앱 배포 준비] Provisioning Profiles 등록](https://ios-development.tistory.com/250)

## 생성하기
1. Mac의 Keychain Access에서 .certSigningRequest 파일 생성하기
    1. 가장 먼저 Mac의 Keychain Acess를 실행합니다. 이 앱은 이 과정에서 계속해 사용되니 켜 두는 것이 편리합니다.
    2. Keychain Acess → certificates Assistant → Request a Certificates From a Certificate Authority… 를 선택해 접근합니다.
        
        <img width="689" alt="Screenshot 2023-03-05 at 18 55 20" src="https://user-images.githubusercontent.com/94854258/224022651-dc21c60c-b82d-49b2-aa39-7cb9cb4fe563.png">

        
    3. User Email Adress에 개발자 계정이 연결된 Email 주소를 입력합니다. 그리고  Request is: 항목에서 Saved to disk를 선택하고 Continue를 클릭하면 .certSigningRequest 파일이 생성됩니다. 이후 파일명과 위치를 지정해주면 저장됩니다.
        
        <img width="728" alt="Screenshot 2023-03-05 at 19 00 06" src="https://user-images.githubusercontent.com/94854258/224022713-ca60892c-836d-47c7-81bc-843cd1873b95.png">
        
2. Apple Developer 웹사이트에서 Certificate 생성하기
    1. Apple Developer 웹사이트의 Account를 선택한 뒤 Certificates에 들어갑니다.
        
        <img width="1271" alt="Screenshot 2023-03-05 at 19 04 03" src="https://user-images.githubusercontent.com/94854258/224023014-c0947fb5-c2fb-4ad3-829d-ecd421335464.png">
        
        
    2. Certificates 오른쪽의 + 버튼을 눌러 Certificates를 생성하는 창으로 들어갑니다. 
    3. 다양한 종류의 Certificates 중 용도에 맞는 유형을 선택합니다. 일반적인 경우 개발 중 실기기 빌드를 위해서는 ‘Apple Development’ 또는 ‘iOS App Development’을, App Store 배포를 위해서는 ‘Apple Distribution’ 또는 ‘iOS App Distribution’을 선택하면 됩니다.
    4. 다음 화면에서 Mac에 생성해두었던 Certificate Signing Request를 업도르 하면 Certificate가 생성됩니다.
    5. 생성된 Certificate를 다운받습니다.
    6. 생성 가능한 Certificate의 개수에는 제한이 있으므로 많은 Certificate를 생성하는 것은 주의해야 합니다. 복수의 프로젝트에 단일 Certificate을 적용할 수 있습니다.
3. Mac에 Certificate 적용하기
    1. 다운로드받은 Certificate를 더블클릭해 실행합니다 (.cer 확장자) 
    2. Keychain Access에 새로운 Public Key 및 Private Key가 생성되었는지 확인합니다.
    3. Private Key 내부에 자신의 Apple Development 계정과 연결된 Certificate가 들어 있어야 합니다.
        
        <img width="1007" alt="Screenshot 2023-03-09 at 18 15 37" src="https://user-images.githubusercontent.com/94854258/224023147-ad12ca2f-0108-4e88-ad7e-60ab82b758cc.png">
        
4. Apple Developer 웹사이트에서 Provisioning Profile 생성하기
    1. Apple Developer 웹사이트의 Account를 선택한 뒤 Profiles에 ****들어갑니다.****
    2. 좌상단의 + 버튼을 눌러 새로운 Profile 생성을 시작합니다.
    3. 아래와 같은 목록에서 용도를 선택해 Profile을 생성합니다. Development 섹션은 개발 중 실기기 빌드 시 사용되는 프로파일을 생성해줍니다. Distribution 섹션은 App Store 등록을 위한 Archive 과정에서 사용되는 프로파일을 생성해줍니다. 일반적인 iOS 앱의 경우 iOS App Development 및 App Store Profile을 생성하면 됩니다.
        
        <img width="993" alt="Screenshot 2023-03-09 at 18 19 26" src="https://user-images.githubusercontent.com/94854258/224023227-2c5fd6b4-da4c-4934-97cc-7fe61f501b84.png">
        
    4. iOS App Development를 선택했을 경우 Profile이 필요한 App ID를 선택하고 Continue를 눌러 계속합니다. 만약 App Group으로 묶인 Extension 등에 사용되는 App ID가 있다면 해당 App ID만큼 Profile을 생성해 주세요.
        
        <img width="991" alt="Screenshot 2023-03-09 at 18 24 13" src="https://user-images.githubusercontent.com/94854258/224023408-61fc2de5-5f87-4197-ba1c-578e63e36cc4.png">
        
    5. 어떤 Certificate와 맞는 Profile을 생성할 것인지 선택해주세요. 일반적인 경우, 하나의 Certificate가 표시됩니다.
        
        <img width="987" alt="Screenshot 2023-03-09 at 18 28 11" src="https://user-images.githubusercontent.com/94854258/224023465-8ce4ea5f-c431-4cf9-aa1d-2941b44e0709.png">

    6. 해당 Profile이 어떤 기기들에서의 Build를 지원할 것인지 선택합니다. 여기에 기기가 표시되기 위해서는 먼저 개발자 계정에 Devices가 등록되어 있어야 합니다. 팀원이 있다면 팀원의 기기도 함께 기기 항목에 추가 후 선택해야 합니다.
        
        <img width="993" alt="Screenshot 2023-03-09 at 18 28 58" src="https://user-images.githubusercontent.com/94854258/224023718-3f66829b-b526-46aa-ac2d-82e81672ddf8.png">

    7. Profile의 이름을 설정해주고 Generate한 뒤 다운로드받습니다.
        
        <img width="993" alt="Screenshot 2023-03-09 at 19 28 05" src="https://user-images.githubusercontent.com/94854258/224023757-a3146c14-b58b-4a2e-b7b1-897a68970c2d.png">
        
    8. App Store용 Profile도 같은 과정을 통해 생성해야 합니다.
5. 프로젝트에 Provisioning Profile 적용하기
    1. Profile을 적용할 앱 프로젝트를 열어서 프로젝트 파일의 Target의 Signing&Capabilities에 들어갑니다.
    2. Signing 섹션의 Provisioning Profile을 눌러 다운로드 받은 Profile을 import해줍니다.
    3. 실기기 빌드 성공!

## 팀원에게 공유하기

- 팀원들에게 해당 환경을 공유하기 위해서는 Provisioning Profile 뿐 아니라 Certificate도 함께 공유해야 합니다.
- Certificate는 Private Key를 .p12 확장자 형태의 파일로 공유하면 됩니다.
- Profile은 다운로드 받은 파일을 그대로 팀원과 공유하면 됩니다.
 
1. Mac의 Keychain Acess에서 .p12파일 생성하기
    1. 팀원들에게 공유할 private Key를 우클릭 선택한 뒤 Export합니다.
    2. 생성된 .p12 확장자를 가진 파일을 팀원들과 공유하면 됩니다.
2. 공유받은 .p12파일 설치하기
    1. 생성자가 공유한 .p12파일을 더블클릭해서 실행합니다.
    2. Keychain Access의 Keys에 생성자의 이름으로 된 Private Key가 성공적으로 공유되었는지 확인합니다. Private Key 내부에 Certificate가 있는지도 확인합니다.
    3. Xcode Project에서 위와 같은 절차로 Certificate와 Provisioning을 설치하면 실기기 빌드가 가능합니다.

**Comment by JIWON1923** [comment]
공유 감사해요! 다른 프로젝트 진행할 때 참고할 수 있겠네요👍
